cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -1553,221 +1057,325 @@ TEST_F(TestCast, UnsupportedTargetType) {
+  auto list_int64 = std::make_shared<ListArray>(
+      list(int64()), list_int32->length(), list_int32->value_offsets(),
+      *Cast(*list_int32->values(), int64()), list_int32->null_bitmap());
+  ASSERT_OK(list_int64->ValidateFull());
 
-TEST_F(TestCast, ListToPrimitive) {
-  auto from_int = ArrayFromJSON(list(int8()), "[[1, 2], [3, 4]]");
-  auto from_binary = ArrayFromJSON(list(binary()), "[[\"1\", \"2\"], [\"3\", \"4\"]]");
+  auto list_float32 = std::make_shared<ListArray>(
+      list(float32()), list_int32->length(), list_int32->value_offsets(),
+      *Cast(*list_int32->values(), float32()), list_int32->null_bitmap());
+  ASSERT_OK(list_float32->ValidateFull());
 
-  ASSERT_RAISES(NotImplemented, Cast(*from_int, uint8()));
-  ASSERT_RAISES(NotImplemented, Cast(*from_binary, utf8()));
-}
+  CheckCast(list_int32, list_float32);
+  CheckCast(list_float32, list_int64);
+  CheckCast(list_int64, list_float32);
 
-TEST_F(TestCast, ListToList) {
-  CastOptions options;
-  std::shared_ptr<Array> offsets;
-
-  std::vector<int32_t> offsets_values = {0, 1, 2, 5, 7, 7, 8, 10};
-  std::vector<bool> offsets_is_valid = {true, true, true, true, false, true, true, true};
-  ArrayFromVector<Int32Type>(offsets_is_valid, offsets_values, &offsets);
-
-  std::shared_ptr<Array> int32_plain_array =
-      TestBase::MakeRandomArray<typename TypeTraits<Int32Type>::ArrayType>(10, 2);
-  ASSERT_OK_AND_ASSIGN(auto int32_list_array,
-                       ListArray::FromArrays(*offsets, *int32_plain_array, pool_));
-
-  ASSERT_OK_AND_ASSIGN(std::shared_ptr<Array> int64_plain_array,
-                       Cast(*int32_plain_array, int64(), options));
-  ASSERT_OK_AND_ASSIGN(auto int64_list_array,
-                       ListArray::FromArrays(*offsets, *int64_plain_array, pool_));
-
-  ASSERT_OK_AND_ASSIGN(std::shared_ptr<Array> float64_plain_array,
-                       Cast(*int32_plain_array, float64(), options));
-  ASSERT_OK_AND_ASSIGN(auto float64_list_array,
-                       ListArray::FromArrays(*offsets, *float64_plain_array, pool_));
-
-  CheckPass(*int32_list_array, *int64_list_array, int64_list_array->type(), options,
-            /*check_scalar=*/false);
-  CheckPass(*int32_list_array, *float64_list_array, float64_list_array->type(), options,
-            /*check_scalar=*/false);
-  CheckPass(*int64_list_array, *int32_list_array, int32_list_array->type(), options,
-            /*check_scalar=*/false);
-  CheckPass(*int64_list_array, *float64_list_array, float64_list_array->type(), options,
-            /*check_scalar=*/false);
-
-  options.allow_float_truncate = true;
-  CheckPass(*float64_list_array, *int32_list_array, int32_list_array->type(), options,
-            /*check_scalar=*/false);
-  CheckPass(*float64_list_array, *int64_list_array, int64_list_array->type(), options,
-            /*check_scalar=*/false);
+  CheckCast(list_int32, list_int64);
+  CheckCast(list_float32, list_int32);
+  CheckCast(list_int64, list_int32);
 }
```

Comment:
Can we add a simple test for `CastOptions` pass-through? e.g. a cast from `list<int32>` to `list<int16>` that would trigger overflow.

Reply:
will do